File: filter.js
/**
 * Returns a new array containing objects that match the filter
 * @param objects        List of JS objects that should be filtered
 * @param attr           The attribute list of the object to check against
 * @param grouping       As built by the Terminator.
 */
// TODO: Filter needs to handle a grouping
var objectFilter = function(objects, attr, grouping) {
 // console.log("items: " + JSON.stringify(grouping.items));
 // console.log("grouping: " + JSON.stringify(grouping) + " type: " + grouping.type);
 // console.log("number of objects:" + objects.length);
  // If no filters, then return just the object
  var filteredObjects = [];
  // console.log("type is string: " + grouping);
  if (typeof grouping == 'string') {
    return _.filter(objects, function(object) {
      return _.contains(object[attr], grouping);
    });
  } else if (grouping.items.length === 0) {
    // console.log("length of items is zero");
    return objects
  }
  if (grouping.type === "AND") {
    // console.log("type is AND");
    var newObjects = objects.slice(0);
    for (var x = 0; x < grouping.items.length; x++) {
      newObjects = _.intersection(newObjects, objectFilter(objects, attr, grouping.items[x]));
    }
    return newObjects;
  } else if (grouping.type === "OR") {
    // console.log("TYPE IS OR");
    for (var y = 0; y < grouping.items.length; y++) {
      filteredObjects = _.union(filteredObjects, objectFilter(objects, attr, grouping.items[y]));
    }
    return filteredObjects;
  } else {
    console.log("SHOULD NOT GET HERE");
    return objects;
  }
};

/*
NOTES
1) It might be useful to parse booleanQuery by another helper method
2) This might be best solved by recursively diving into booleanQuery
    If that's the case, tell Chris.  He can throw together a Set class
    like what's in Java so that you don't have to worry about uniquely 
    merging two arrays.

*/

File: filter.py
 # @param objects        List of JS objects that should be filtered
 # @param attr           The attribute list of the object to check against
 # @param grouping       As built by the Terminator.
import terminator

#print "inside filter"

def objectFilter(objects, attr, string):
  return filterObjects(objects, attr, terminator.terminator(string))

def filterObjects(objects, attr, grouping):
  #print "New grouping: " , grouping, " type: " , type(grouping)
  filteredObjects = [];
  if type(grouping) != type({}):
    #print "grouping is string"
    return [obj for obj in objects if grouping.lower() in map(lambda x: x.lower(), obj[attr])]
  elif len(grouping["items"])==0:
    #print "length is 0"
    return objects
  if grouping["type"] == "AND":
    #print "type is AND"
    newObjects = list(objects);
    for i in xrange(len(grouping["items"])):
      intersect = [val for val in newObjects if val in filterObjects(objects, attr, grouping["items"][i])]
      newObjects = list(intersect)
    return newObjects;
  elif grouping["type"] == "OR":
    #print "type is OR"
    for x in xrange(len(grouping["items"])):
      filteredObjects.extend(filterObjects(objects, attr, grouping["items"][x]))
    return filteredObjects
  return filteredObjects
